phpseclib/Crypt/Base.php

```diff
@@ -1179,16 +1150,113 @@ function disableContinuousBuffer()
 
         $this->continuousBuffer = false;
         $this->changed = true;
+
+        $this->_setEngine();
     }
 
     /**
-     * Encrypts a block
+     * Test for engine validity
+     *
+     * @see Crypt_Base::Crypt_Base()
+     * @param Integer $engine
+     * @access public
+     * @return Boolean
+     */
+    function isValidEngine($engine)
+    {
+        switch ($engine) {
+            case CRYPT_ENGINE_MCRYPT:
+                return $this->cipher_name_mcrypt &&
+                       extension_loaded('mcrypt') &&
+                       in_array($this->cipher_name_mcrypt, mcrypt_list_algorithms());
+            case CRYPT_ENGINE_INTERNAL:
+                return true;
+        }
+    }
+
+    /**
+     * Sets the preferred crypt engine
+     *
+     * Currently, $engine could be:
+     *
+     * - CRYPT_ENGINE_MCRYPT   [fast]
+     *
+     * - CRYPT_ENGINE_INTERNAL [slow]
+     *
+     * If the preferred crypt engine is not available the fastest available one will be used
      *
-     * Note: Must extend by the child Crypt_* class
+     * @see Crypt_Base::Crypt_Base()
+     * @param Integer $engine
+     * @access public
+     */
+    function setPreferredEngine($engine)
+    {
+        switch ($engine) {
+            //case CRYPT_ENGINE_MCRYPT:
+            case CRYPT_ENGINE_INTERNAL:
+                $this->preferredEngine = $engine;
+                break;
+            default:
+                $this->preferredEngine = CRYPT_ENGINE_MCRYPT;
+        }
+
+        $this->_setEngine();
+    }
+
+    /**
+     * Returns the engine currently being utilized
+     *
+     * @see Crypt_Base::_setEngine()
+     * @access public
+     */
+    function getEngine()
+    {
+        return $this->engine;
+    }
+
+    /**
+     * Sets the engine as appropriate
+     *
+     * @see Crypt_Base::Crypt_Base()
+     * @access private
+     */
+    function _setEngine()
+    {
+        switch (true) {
+            case $this->isValidEngine($this->preferredEngine):
+                $this->engine = $this->preferredEngine;
+                break;
+            case $this->isValidEngine(CRYPT_ENGINE_MCRYPT):
```

Comment:
I can do that. Altho we'd need to do `$this->engine = false` before all of that lest something like this happen.

For the initial key / IV length mcrypt is a valid engine, so `$this->engine` gets set to CRYPT_ENGINE_MCRYPT. Then you change the key length to one for which only the internal engine is valid. `$this->engine` is never changed to internal because `$this->engine` is never false (or NULL or 0).